File: cpp/medium/Combination_Sum.cpp
class Solution {
public:
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        vector<vector<int>> answer;
        vector<int> combination;
        layer(0, answer, candidates, target, combination);
        return answer;
    }
    void layer(int j, vector<vector<int>>& answer, vector<int>& candidates, int target, vector<int>& combination) {
        for (int i = j; i < candidates.size(); ++i){
            combination.push_back(candidates[i]);
            target-=candidates[i];
            // for (int n = 0; n < combination.size(); ++n){
            //     cout << combination[n] << ",";
            // }
            // cout << "; ";
            // cout << i << endl;
            if (target==0) {
                answer.push_back(combination);
                combination.pop_back();
                target+=candidates[i];
            }
            else if (target>0) {
                // vector<int> sub_candidates(candidates.begin()+i, candidates.end());
                layer(i, answer, candidates, target, combination);
                combination.pop_back();
                target+=candidates[i];
            }
            else{
                combination.pop_back();
                target+=candidates[i];
            }
        }
    }
};

File: python/medium/Combination_Sum.py
class Solution:
    def __init__(self):
        self.candidates = None
        self.answer = []
    
    def candidate_sum(self, combinations: List[int], target: int) -> None:
        target = target - combinations[-1]
        if target == 0:
            if sorted(combinations) not in self.answer:
                self.answer.append(sorted(combinations))
        elif target < 0:
            pass
        else:
            for digit in self.candidates:
                self.candidate_sum(combinations+[digit], target)
        
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        self.candidates = candidates
        for digit in candidates:
            self.candidate_sum([digit], target)
        return self.answer


# class Solution:
#     def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
#         answer = []
#         candidates = sorted(candidates)
#         def candidate_sum(candidates: List[int], combinations: List[int], target: int) -> None:
#             target = target - combinations[-1]
#             if target == 0:
#                 if sorted(combinations) not in answer:
#                     answer.append(sorted(combinations))
#             elif target < 0:
#                 pass
#             else:
#                 for digit in candidates:
#                     if digit > target:
#                         break
#                     candidate_sum(candidates, combinations+[digit], target)
#         for index in range(len(candidates)):
#             candidate_sum(candidates[index:], [candidates[index]], target)
#         return answer
